File: 27. XGBoost.R
## XG boosting ##


# libraries ---------------------------------------------------------------

library(tidyverse)
library(caTools)
library(xgboost)

# read data ---------------------------------------------------------------

datos <- readr::read_csv("data/Churn_Modelling.csv")

datos <- datos %>% 
  dplyr::select(-(1:3)) %>% 
  dplyr::mutate(Geography = factor(Geography),
                Gender = factor(Gender))

str(datos)

datos_xg <- model.matrix(Exited ~ ., data = datos)[,-1]

head(datos_xg)

# Entrenamiento y validación ----------------------------------------------

set.seed(123)

split <- caTools::sample.split(datos_xg[,1], SplitRatio = 0.75)

entrenamiento <- datos_xg %>% 
  subset(split == T)

ent_etiqueta <- datos %>% 
  subset(split == T) %>% 
  .$Exited

validacion <- datos_xg %>% 
  subset(split == F)

val_etiqueta <- datos %>% 
  subset(split == F) %>% 
  .$Exited


# ajustar clasificador ----------------------------------------------------

clasificador <- xgboost::xgboost(data = entrenamiento,
                                 nrounds = 200,
                                 label = ent_etiqueta, 
                                 verbose = TRUE, 
                                 objective = "binary:logistic")

y_pred <- as.numeric(predict(clasificador, validacion)>0.5)

# matriz confusion --------------------------------------------------------

table(val_etiqueta, y_pred)

sum(diag(table(val_etiqueta, y_pred)))/sum(table(val_etiqueta, y_pred))

# validación cruzada ------------------------------------------------------

set.seed(123)

folds <- caret::createFolds(datos$Exited)

precisiones <- folds %>% 
  purrr::map_dbl(function(x){
    
    entrenamiento <- datos[-x, ]
    validacion <- datos[x, ]
    
    modelo <- xgboost::xgboost(data = model.matrix(Exited ~ ., 
                                                   data = entrenamiento)[,-1],
                               nrounds = 200,
                               label = entrenamiento$Exited, 
                               verbose = FALSE, 
                               objective = "binary:logistic")
    
    prediccion <- as.numeric(
      predict(clasificador, 
              model.matrix(Exited ~ ., 
                           data = validacion)[,-1])>0.5)
    
    cm <- table(validacion$Exited, prediccion)
    
    sum(diag(cm))/sum(cm)
    
  })


mean(precisiones); sd(precisiones)

File: 27. XGBoost.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 30 15:00:03 2022

@author: Admin
"""

## xgboost ##

# =============================================================================
# Libraries
# =============================================================================

import numpy as np
import pandas as pd
import sklearn.preprocessing as sklp
import sklearn.model_selection as sklm
import sklearn.compose as sklc
import sklearn.metrics as sklmetric

# =============================================================================
# Read data
# =============================================================================

churn = pd.read_csv("data\Churn_Modelling.csv")

x = churn.iloc[:, 3:13].values
y = churn.iloc[:, 13].values


# =============================================================================
# Datos categóricos
# =============================================================================

label_encoder_x1 = sklp.LabelEncoder()
x[:, 1] = label_encoder_x1.fit_transform(x[:, 1])

label_encoder_x2 = sklp.LabelEncoder()
x[:, 2] = label_encoder_x2.fit_transform(x[:, 2])

ct = sklc.ColumnTransformer(
    [("one_hot_encoder", sklp.OneHotEncoder(categories = "auto"), [1])],
    remainder = "passthrough")

x = np.array(ct.fit_transform(x), dtype = np.float64)

x = x[:,1:]

# =============================================================================
# División de entrenamiento y test
# =============================================================================

x_train, x_test, y_train, y_test = sklm.train_test_split(x, y, test_size= 0.25,
                                                         random_state = 0)

# =============================================================================
# Ajustar XGBoost
# =============================================================================

from xgboost import XGBClassifier as XGB

clasificador = XGB()
clasificador.fit(x_train, y_train)

# =============================================================================
# Predicción de resultados 
# =============================================================================

y_pred = clasificador.predict(x_test)

# =============================================================================
# Matriz de confusion
# =============================================================================

cm = sklmetric.confusion_matrix(y_test, y_pred)
cm

np.sum(np.diag(cm))/np.sum(cm)

# =============================================================================
# Validacion cruzada
# =============================================================================

from sklearn.model_selection import cross_val_score as CV

precisiones =  CV(estimator = clasificador, X = x_train, y = y_train, cv = 10)

precisiones.mean()
precisiones.std()
